### PR TITLE
[AEA-557] Fix for failing tests

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes
Using pexpect to check subprocesses output

## Fixes

cli/launch test updated with pexpect.
github ci better reporting: -rfE flag
fixes for CWD. auto root dir and cwd restore check.
stub connection on mac os x uses kqueue cause issue with mulithreading and several agents.

## Types of changes

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [X] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [X] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [X] Lint and unit tests pass locally with my changes
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] I have checked that code coverage does not decrease.
- [X] I have checked that the documentation about the `aea cli` tool works
- [X] I have added necessary documentation (if appropriate)
- [X] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: .github/workflows/workflow.yml [hunk 1]
```diff
@@ -70,11 +70,11 @@ jobs:
           pip install pipenv
           pip install tox
       - name: Integration tests
-        run: tox -e py3.7 -- -m integration
+        run: tox -e py3.7 -- -rfE -m integration
```
Comment (DavidMinarsch) [inline .github/workflows/workflow.yml:73]: What does `-rfE` do?

File: .github/workflows/workflow.yml [hunk 2]
```diff
@@ -70,11 +70,11 @@ jobs:
...
 
   platform_checks:
     runs-on: ${{ matrix.os }}
-
+    continue-on-error: true
```
Comment (DavidMinarsch) [inline .github/workflows/workflow.yml:77]: I think we should remove this again to avoid burning too many minutes.

File: aea/cli/launch.py [hunk 3]
```diff
@@ -121,7 +122,11 @@ def _launch_threads(click_context: click.Context, agents: List[Path]):
...
-        _launch_subprocesses(click_context, agents_directories)
+    if multithreaded:
+        failed = _launch_threads(click_context, agents_directories)
     else:
-        _launch_threads(click_context, agents_directories)
+        failed = _launch_subprocesses(click_context, agents_directories)
+    logger.debug(f"Exit cli. code: {failed}")
+    sys.stdout.flush()
+    sys.stderr.flush()
+    sys.exit(failed)
```
Comment (DavidMinarsch) [inline aea/cli/launch.py:132]: Why is the flush needed? Just curious...

File: tests/common/utils.py [hunk 4]
```diff
@@ -249,3 +251,27 @@ def teardown(self):
...
+    """
+    Chdir to ROOT DIR and return back during tests.
+
+    Decorator.
+
+    :param fn: function to decorate
+
+    :return: wrapped function
+    """
+    # pydocstyle: igonre # case confilct with black
```
Comment (DavidMinarsch) [inline tests/common/utils.py:266]: typo

File: tests/test_cli/test_gui.py [hunk 5]
```diff
@@ -16,56 +16,59 @@
...
 )
 
 
-class TestGui:
+class TestGui(TestCase):
     """Test that the command 'aea gui' works as expected."""
 
-    @classmethod
-    def setup_class(cls):
+    def setUp(self):
```
Comment (DavidMinarsch) [inline tests/test_cli/test_gui.py:44]: is `setUp` the setup method of `TestCase`?

File: tests/test_cli/test_gui.py [hunk 6]
```diff
@@ -16,56 +16,59 @@
...
-        assert tcpping("127.0.0.1", 8080)
+        self.proc = PexpectSpawn(  # nosec
+            sys.executable,
+            ["-m", "aea.cli", "-v", "DEBUG", "gui"],
+            encoding="utf-8",
+            logfile=sys.stdout,
+        )
+        self.proc.expect_exact(["Running on http://"], timeout=20)
+
+        assert tcpping("localhost", 8080)
```
Comment (DavidMinarsch) [inline tests/test_cli/test_gui.py:68]: Note, this must be "127.0.0.1" instead of "localhost"..more reliable

File: tests/test_cli/test_launch.py [hunk 7]
```diff
@@ -16,32 +16,76 @@
...
 logger = logging.getLogger(__name__)
 
 
-class TestLaunch:
-    """Test that the command 'aea launch <agent_name>' works as expected."""
+class BaseLaunchTestCase:
+    """Base Test case for launch tests."""
+
+    @classmethod
+    def setUpClass(cls) -> None:
```
Comment (DavidMinarsch) [inline tests/test_cli/test_launch.py:51]: I think we should stick to `snack_case`

File: tox.ini [hunk 8]
```diff
@@ -24,12 +24,13 @@ deps =
...
     SQLAlchemy==1.3.16
     pynacl==1.3.0
+    pexpect==4.8.0
 
 commands =
     pip install git+https://github.com/pytoolz/cytoolz.git#egg=cytoolz==0.10.1.dev0
     pip install -e .[all]
     pip install -i https://test.pypi.org/simple/ fetch-p2p-api==0.0.2
-    pytest --doctest-modules aea packages/fetchai/protocols packages/fetchai/connections tests/ --cov-report=html --cov-report=xml --cov-report=term --cov=aea --cov=packages/fetchai/protocols --cov=packages/fetchai/connections {posargs}
+    pytest -rf --doctest-modules aea packages/fetchai/protocols packages/fetchai/connections tests/ --cov-report=html --cov-report=xml --cov-report=term --cov=aea --cov=packages/fetchai/protocols --cov=packages/fetchai/connections {posargs}
```
Comment (DavidMinarsch) [inline tox.ini:33]: What does `-rf` do?

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (solarw) [reply .github/workflows/workflow.yml:73]: it shows final report on failed and error tests. make it a bit easier to know what failed

File: .github/workflows/workflow.yml (hunk 2 above)
Comment (solarw) [reply .github/workflows/workflow.yml:77]: removed. used only for testing.

File: tox.ini (hunk 8 above)
Comment (solarw) [reply tox.ini:33]: reporting: provides list of failed tests at the end of output. just make it easier to see what tets failed

File: tests/test_cli/test_gui.py (hunk 5 above)
Comment (solarw) [reply tests/test_cli/test_gui.py:44]: yes, fixed. changed to pytest's style

File: tests/test_cli/test_gui.py (hunk 6 above)
Comment (solarw) [reply tests/test_cli/test_gui.py:68]: done

File: tests/test_cli/test_launch.py (hunk 7 above)
Comment (solarw) [reply tests/test_cli/test_launch.py:51]: it's python unit test naming:
https://docs.python.org/3/library/unittest.html
but will switch to pytest's style.

File: aea/cli/launch.py (hunk 3 above)
Comment (solarw) [reply aea/cli/launch.py:132]: I suppose there was a problem with process stdout flush on close, so pexpect can not get final lines. but solved. flush removed

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (solarw) [reply .github/workflows/workflow.yml:73]: removed, cause duplicate in tox.ini

File: tests/test_cli/test_launch.py (hunk 7 above)
Comment (DavidMinarsch) [reply tests/test_cli/test_launch.py]: hmm ok

File: .github/workflows/workflow.yml (hunk 1 above)
Comment (DavidMinarsch) [reply .github/workflows/workflow.yml:73]: Still here ...

Comment (solarw) [reply .github/workflows/workflow.yml]: fixed

File: tests/common/utils.py (hunk 4 above)
Comment (solarw) [reply tests/common/utils.py]: fixed

**Pull request merged**